In [3]:
import numpy as np 
import pandas as pd 
from pandas_profiling import ProfileReport
import matplotlib.pyplot as plt
import seaborn as sns

import os
for dirname, _, filenames in os.walk('/kaggle/input/book-recommendation-dataset'):
    for filename in filenames:
        print(os.path.join(dirname, filename))
        
        
import warnings
warnings.filterwarnings("ignore")


In [8]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

In [9]:
!kaggle datasets download -d arashnic/book-recommendation-dataset

100% 23.8M/23.8M [00:01<00:00, 24.9MB/s]
100% 23.8M/23.8M [00:01<00:00, 14.6MB/s]


In [10]:
!unzip book-recommendation-dataset.zip

Archive:  book-recommendation-dataset.zip
  inflating: Books.csv               
  inflating: Ratings.csv             
  inflating: Users.csv               


In [11]:
books = pd.read_csv("/content/Books.csv")
users = pd.read_csv("/content/Users.csv")
ratings = pd.read_csv("/content/Ratings.csv")

In [12]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [13]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [14]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [15]:
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 8)
(1149780, 3)
(278858, 3)


In [16]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [17]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [18]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [19]:
print(books.duplicated().sum())
print(users.duplicated().sum())
print(ratings.duplicated().sum())

0
0
0


In [20]:
ratings_with_name = ratings.merge(books,on='ISBN')

In [21]:
ratings_with_name.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [22]:
# num_rating_df=ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
# num_rating_df.rename(columns={'Book-Rating':"num_rating"},inplace=True)
# num_rating_df.head()

In [23]:
ratings_with_name.drop(columns=["ISBN","Image-URL-S","Image-URL-M"],axis=1,inplace=True)

In [24]:
complete_df = ratings_with_name.merge(users.drop("Age", axis=1), on="User-ID")
complete_df.head()

User-ID  Book-Rating                                        Book-Title  \
0   276725            0                              Flesh Tones: A Novel   
1     2313            5                              Flesh Tones: A Novel   
2     2313            9     Ender's Game (Ender Wiggins Saga (Paperback))   
3     2313            8             In Cold Blood (Vintage International)   
4     2313            9  Divine Secrets of the Ya-Ya Sisterhood : A Novel   

        Book-Author Year-Of-Publication         Publisher  \
0        M. J. Rose                2002  Ballantine Books   
1        M. J. Rose                2002  Ballantine Books   
2  Orson Scott Card                1986         Tor Books   
3     TRUMAN CAPOTE                1994           Vintage   
4     Rebecca Wells                1996     HarperCollins   

                                         Image-URL-L               Location  
0  http://images.amazon.com/images/P/034545104X.0...      tyler, texas, usa  
1  http://images.amazon.com/images/P/034545104X.0...  cincinnati, ohio, usa  
2  http://images.amazon.com/images/P/0812533550.0...  cincinnati, ohio, usa  
3  http://images.amazon.com/images/P/0679745580.0...  cincinnati, ohio, usa  
4  http://images.amazon.com/images/P/0060173289.0...  cincinnati, ohio, usa

In [26]:
complete_df['Location'] = complete_df['Location'].str.split(',').str[-1].str.strip()

In [27]:
complete_df.head()

User-ID  Book-Rating                                        Book-Title  \
0   276725            0                              Flesh Tones: A Novel   
1     2313            5                              Flesh Tones: A Novel   
2     2313            9     Ender's Game (Ender Wiggins Saga (Paperback))   
3     2313            8             In Cold Blood (Vintage International)   
4     2313            9  Divine Secrets of the Ya-Ya Sisterhood : A Novel   

        Book-Author Year-Of-Publication         Publisher  \
0        M. J. Rose                2002  Ballantine Books   
1        M. J. Rose                2002  Ballantine Books   
2  Orson Scott Card                1986         Tor Books   
3     TRUMAN CAPOTE                1994           Vintage   
4     Rebecca Wells                1996     HarperCollins   

                                         Image-URL-L Location  
0  http://images.amazon.com/images/P/034545104X.0...      usa  
1  http://images.amazon.com/images/P/034545104X.0...      usa  
2  http://images.amazon.com/images/P/0812533550.0...      usa  
3  http://images.amazon.com/images/P/0679745580.0...      usa  
4  http://images.amazon.com/images/P/0060173289.0...      usa

In [28]:
complete_df[complete_df['Location']=='singapore'].head()

User-ID  Book-Rating  \
352041   128461            9   
524323   120565            9   
524324   120565            8   
524325   120565            9   
524326   120565            9   

                                               Book-Title        Book-Author  \
352041                                Saving Private Ryan  Max Allan Collins   
524323  Harold and the Purple Crayon 50th Anniversary ...   Crockett Johnson   
524324  The Venetian's Wife: A Strangely Sensual Tale ...       Nick Bantock   
524325                                      The Testament       John Grisham   
524326                                A Multitude of Sins       RICHARD FORD   

       Year-Of-Publication          Publisher  \
352041                1998  Penguin Books Ltd   
524323                1981       HarperTrophy   
524324                1996    Chronicle Books   
524325                1999   Trafalgar Square   
524326                2002              Knopf   

                                              Image-URL-L   Location  
352041  http://images.amazon.com/images/P/0140279091.0...  singapore  
524323  http://images.amazon.com/images/P/0064430227.0...  singapore  
524324  http://images.amazon.com/images/P/0811811409.0...  singapore  
524325  http://images.amazon.com/images/P/0099245027.0...  singapore  
524326  http://images.amazon.com/images/P/0375412123.0...  singapore

In [29]:
num_rating_df = complete_df.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating': 'num_ratings'}, inplace=True)
num_rating_df.head(10)

Book-Title  num_ratings
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1
2               Apple Magic (The Collector's series)            1
3   Ask Lily (Young Women of Faith: Lily Series, ...            1
4   Beyond IBM: Leadership Marketing and Finance ...            1
5   Clifford Visita El Hospital (Clifford El Gran...            1
6                                       Dark Justice            1
7                                           Deceived            2
8   Earth Prayers From around the World: 365 Pray...           10
9   Final Fantasy Anthology: Official Strategy Gu...            4

In [30]:
avg_rating_df = complete_df.groupby('Book-Title').mean()['Book-Rating'].reset_index()
avg_rating_df.rename(columns={'Book-Rating': 'avg_ratings'}, inplace=True)
avg_rating_df.head(10)

Book-Title  avg_ratings
0   A Light in the Storm: The Civil War Diary of ...         2.25
1                              Always Have Popsicles         0.00
2               Apple Magic (The Collector's series)         0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...         8.00
4   Beyond IBM: Leadership Marketing and Finance ...         0.00
5   Clifford Visita El Hospital (Clifford El Gran...         0.00
6                                       Dark Justice        10.00
7                                           Deceived         0.00
8   Earth Prayers From around the World: 365 Pray...         5.00
9   Final Fantasy Anthology: Official Strategy Gu...         5.00

In [31]:
popularity_df = num_rating_df.merge(avg_rating_df, on='Book-Title')
popularity_df.head(10)

Book-Title  num_ratings  avg_ratings
0   A Light in the Storm: The Civil War Diary of ...            4         2.25
1                              Always Have Popsicles            1         0.00
2               Apple Magic (The Collector's series)            1         0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...            1         8.00
4   Beyond IBM: Leadership Marketing and Finance ...            1         0.00
5   Clifford Visita El Hospital (Clifford El Gran...            1         0.00
6                                       Dark Justice            1        10.00
7                                           Deceived            2         0.00
8   Earth Prayers From around the World: 365 Pray...           10         5.00
9   Final Fantasy Anthology: Official Strategy Gu...            4         5.00

In [32]:
top_50 = popularity_df[popularity_df['num_ratings']>=250].sort_values("avg_ratings",ascending=False).head(50)
top_50.head()

Book-Title  num_ratings  \
80434  Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422       Harry Potter and the Goblet of Fire (Book 4)          387   
80441     Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426  Harry Potter and the Order of the Phoenix (Boo...          347   
80414   Harry Potter and the Chamber of Secrets (Book 2)          556   

       avg_ratings  
80434     5.852804  
80422     5.824289  
80441     5.737410  
80426     5.501441  
80414     5.183453

In [33]:
final_top_50 = top_50.merge(books, on='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Publisher','num_ratings','avg_ratings','Image-URL-M']]

In [34]:
final_top_50['Image-URL-M'][0]

'http://images.amazon.com/images/P/0439136350.01.MZZZZZZZ.jpg'

In [35]:
final_top_50.head()

Book-Title    Book-Author  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)  J. K. Rowling   
3        Harry Potter and the Goblet of Fire (Book 4)  J. K. Rowling   
5      Harry Potter and the Sorcerer's Stone (Book 1)  J. K. Rowling   
9   Harry Potter and the Order of the Phoenix (Boo...  J. K. Rowling   
13   Harry Potter and the Chamber of Secrets (Book 2)  J. K. Rowling   

     Publisher  num_ratings  avg_ratings  \
0   Scholastic          428     5.852804   
3   Scholastic          387     5.824289   
5   Scholastic          278     5.737410   
9   Scholastic          347     5.501441   
13  Scholastic          556     5.183453   

                                          Image-URL-M  
0   http://images.amazon.com/images/P/0439136350.0...  
3   http://images.amazon.com/images/P/0439139597.0...  
5   http://images.amazon.com/images/P/0590353403.0...  
9   http://images.amazon.com/images/P/043935806X.0...  
13  http://images.amazon.com/images/P/0439064872.0...

In [36]:
final_top_50.shape

(50, 6)

### Collaborative Filtering Based Recommender System


In [37]:
x = complete_df.groupby('User-ID').count()['Book-Rating']>200
knowledgable_users = x[x].index

In [38]:
filtered_rating = complete_df[complete_df['User-ID'].isin(knowledgable_users)]

In [39]:
y = filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books = y[y].index

In [40]:
final_ratings =  filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]

In [41]:
pt = final_ratings.pivot_table(index='Book-Title',columns='User-ID'
                          ,values='Book-Rating')
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [42]:
pt.fillna(0,inplace=True)
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [43]:
from sklearn.metrics.pairwise import cosine_similarity 

In [44]:
similarity_score = cosine_similarity(pt)

In [45]:
similarity_score.shape

(706, 706)

In [46]:
def recommend(book_name):
    index = np.where(pt.index==book_name)[0][0]
    similar_books = sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1], reverse=True)[1:6]
    
    data = []
    
    for i in similar_books:
        item = []
        temp_df = books[books['Book-Title'] == pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
    return data

In [47]:
recommend("Message in a Bottle")

[['Nights in Rodanthe',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446531332.01.MZZZZZZZ.jpg'],
 ['The Mulberry Tree',
  'Jude Deveraux',
  'http://images.amazon.com/images/P/0743437640.01.MZZZZZZZ.jpg'],
 ['A Walk to Remember',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446608955.01.MZZZZZZZ.jpg'],
 ["River's End",
  'Nora Roberts',
  'http://images.amazon.com/images/P/0515127833.01.MZZZZZZZ.jpg'],
 ['Nightmares &amp; Dreamscapes',
  'Stephen King',
  'http://images.amazon.com/images/P/0451180232.01.MZZZZZZZ.jpg']]

In [48]:
recommend("Zen and the Art of Motorcycle Maintenance: An Inquiry into Values")

[['Prodigal Summer',
  'Barbara Kingsolver',
  'http://images.amazon.com/images/P/0060199652.01.MZZZZZZZ.jpg'],
 ['Pet Sematary',
  'Stephen King',
  'http://images.amazon.com/images/P/0451162072.01.MZZZZZZZ.jpg'],
 ['Midnight',
  'Dean R. Koontz',
  'http://images.amazon.com/images/P/0425118703.01.MZZZZZZZ.jpg'],
 ["The Hitchhiker's Guide to the Galaxy",
  'Douglas Adams',
  'http://images.amazon.com/images/P/0671461494.01.MZZZZZZZ.jpg'],
 ['A Prayer for Owen Meany',
  'John Irving',
  'http://images.amazon.com/images/P/0345361792.01.MZZZZZZZ.jpg']]

In [49]:
recommend("The Notebook")

[['A Walk to Remember',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446608955.01.MZZZZZZZ.jpg'],
 ['The Rescue',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446610399.01.MZZZZZZZ.jpg'],
 ['One Door Away from Heaven',
  'Dean R. Koontz',
  'http://images.amazon.com/images/P/0553582755.01.MZZZZZZZ.jpg'],
 ['Toxin',
  'Robin Cook',
  'http://images.amazon.com/images/P/0425166619.01.MZZZZZZZ.jpg'],
 ['The Five People You Meet in Heaven',
  'Mitch Albom',
  'http://images.amazon.com/images/P/0786868716.01.MZZZZZZZ.jpg']]

In [50]:
recommend("1984")

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'],
 ['The Hours : A Novel',
  'Michael Cunningham',
  'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]

In [51]:
recommend("Harry Potter and the Prisoner of Azkaban (Book 3)")

[['Harry Potter and the Goblet of Fire (Book 4)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0439139597.01.MZZZZZZZ.jpg'],
 ['Harry Potter and the Chamber of Secrets (Book 2)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0439064872.01.MZZZZZZZ.jpg'],
 ['Harry Potter and the Order of the Phoenix (Book 5)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/043935806X.01.MZZZZZZZ.jpg'],
 ["Harry Potter and the Sorcerer's Stone (Book 1)",
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0590353403.01.MZZZZZZZ.jpg'],
 ["Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))",
  'J. K. Rowling',
  'http://images.amazon.com/images/P/059035342X.01.MZZZZZZZ.jpg']]

In [80]:
complete_df2 = ratings_with_name.merge(users.drop("Age", axis=1), on="User-ID")
complete_df2.head()

User-ID  Book-Rating                                        Book-Title  \
0   276725            0                              Flesh Tones: A Novel   
1     2313            5                              Flesh Tones: A Novel   
2     2313            9     Ender's Game (Ender Wiggins Saga (Paperback))   
3     2313            8             In Cold Blood (Vintage International)   
4     2313            9  Divine Secrets of the Ya-Ya Sisterhood : A Novel   

        Book-Author Year-Of-Publication         Publisher  \
0        M. J. Rose                2002  Ballantine Books   
1        M. J. Rose                2002  Ballantine Books   
2  Orson Scott Card                1986         Tor Books   
3     TRUMAN CAPOTE                1994           Vintage   
4     Rebecca Wells                1996     HarperCollins   

                                         Image-URL-L               Location  
0  http://images.amazon.com/images/P/034545104X.0...      tyler, texas, usa  
1  http://images.amazon.com/images/P/034545104X.0...  cincinnati, ohio, usa  
2  http://images.amazon.com/images/P/0812533550.0...  cincinnati, ohio, usa  
3  http://images.amazon.com/images/P/0679745580.0...  cincinnati, ohio, usa  
4  http://images.amazon.com/images/P/0060173289.0...  cincinnati, ohio, usa

In [89]:
df=complete_df2[complete_df2['Book-Rating']>0]

In [90]:
# counts1 = df['User-ID'].value_counts()
# df= df[df['User-ID'].isin(counts1[counts1 > 200].index)]
# df

In [91]:
# len(df['ISBN'].unique())

In [92]:
# cdf1=df[['UserID','ISBN','Rating']]
# cdf=pd.merge(cdf1,books,on='ISBN')
cdf=df
cdf.head(1)

User-ID  Book-Rating            Book-Title Book-Author Year-Of-Publication  \
1     2313            5  Flesh Tones: A Novel  M. J. Rose                2002   

          Publisher                                        Image-URL-L  \
1  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   

                Location  
1  cincinnati, ohio, usa

In [93]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
cdf['User-ID']=le.fit_transform(cdf['User-ID'])
cdf['Book-Title']=le.fit_transform(cdf['Book-Title'])
# cdf.drop('User-ID',axis=1,inplace=True)
cdf

User-ID  Book-Rating  Book-Title                 Book-Author  \
1            564            5       38514                  M. J. Rose   
2            564            9       34428            Orson Scott Card   
3            564            8       51083               TRUMAN CAPOTE   
4            564            9       29941               Rebecca Wells   
5            564            5      111375  Chitra Banerjee Divakaruni   
...          ...          ...         ...                         ...   
1031130    67485            6       56552                Ellis Peters   
1031131    67485            7       58805           Katherine Neville   
1031132    67514            5       63445                Ludwig Marum   
1031134    67522           10       46089              Angela Johnson   
1031135    67527            8       81785          David Antony Kroft   

        Year-Of-Publication                                   Publisher  \
1                      2002                            Ballantine Books   
2                      1986                                   Tor Books   
3                      1994                                     Vintage   
4                      1996                               HarperCollins   
5                      1998                      Anchor Books/Doubleday   
...                     ...                                         ...   
1031130                2001                              Editions 10/18   
1031131                2002                             Le Cherche Midi   
1031132                1984                                C.F. Müller   
1031134                1998  Simon &amp; Schuster Children's Publishing   
1031135                1985                    HarperCollins Publishers   

                                               Image-URL-L  \
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/0812533550.0...   
3        http://images.amazon.com/images/P/0679745580.0...   
4        http://images.amazon.com/images/P/0060173289.0...   
5        http://images.amazon.com/images/P/0385482388.0...   
...                                                    ...   
1031130  http://images.amazon.com/images/P/2264032960.0...   
1031131  http://images.amazon.com/images/P/2862749796.0...   
1031132  http://images.amazon.com/images/P/3788097000.0...   
1031134  http://images.amazon.com/images/P/0689822294.0...   
1031135  http://images.amazon.com/images/P/0583307841.0...   

                                 Location  
1                   cincinnati, ohio, usa  
2                   cincinnati, ohio, usa  
3                   cincinnati, ohio, usa  
4                   cincinnati, ohio, usa  
5                   cincinnati, ohio, usa  
...                                   ...  
1031130       genève, genève, switzerland  
1031131       genève, genève, switzerland  
1031132      stuttgart, \n/a\"., germany"  
1031134  arlington heights, illinois, usa  
1031135         singapore, n/a, singapore  

[383842 rows x 8 columns]

In [94]:
!pip install surprise

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [95]:
cdf.head(1)

User-ID  Book-Rating  Book-Title Book-Author Year-Of-Publication  \
1      564            5       38514  M. J. Rose                2002   

          Publisher                                        Image-URL-L  \
1  Ballantine Books  http://images.amazon.com/images/P/034545104X.0...   

                Location  
1  cincinnati, ohio, usa

In [97]:
from surprise import Reader
from surprise import Dataset
from surprise.model_selection import cross_validate
from surprise import SVD
reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(cdf[['User-ID','Book-Title','Book-Rating']], reader)
svd = SVD()
cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)


Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.6311  1.6382  1.6355  1.6322  1.6409  1.6356  0.0037  
MAE (testset)     1.2589  1.2613  1.2625  1.2608  1.2642  1.2615  0.0018  
Fit time          10.54   8.89    8.94    9.00    8.79    9.23    0.66    
Test time         0.90    0.77    0.76    0.77    1.28    0.90    0.20    


{'test_rmse': array([1.63105428, 1.63822603, 1.63552369, 1.63222384, 1.6409043 ]),
 'test_mae': array([1.25887297, 1.2613004 , 1.26248751, 1.26078926, 1.26416587]),
 'fit_time': (10.540659427642822,
  8.889668941497803,
  8.940218925476074,
  9.004787683486938,
  8.786712169647217),
 'test_time': (0.9002792835235596,
  0.7735412120819092,
  0.7587342262268066,
  0.7713830471038818,
  1.27986741065979)}


|  | RMSE(test set) | MAE(test set) |
| ----------| : ----------| : -------|
| Mean | 1.6356 | 1.2615 |
| Std | 0.0037  | 0.0018 |
